### ISSUE TITLE
Distributed cache for scaling

### ISSUE THREAD

**Issue by tananaev** [traccar/traccar#4118]
@Abyss777, I'm experimenting with replacing our in-memory caching with JCache. Do you remember why we update existing devices in cache instead of replacing them with completely new instances of the objects?

**Comment by Abyss777** [comment]
I'm not sure, but I think because of some out of sync during pipeline. Especially with `positionId`. Also `Device` object updated frequently: `positionId`, `lastUpdate`, `status` etc. I believe with every position.

**Comment by tananaev** [comment]
I understand that there might be some edge cases for devices, but for example in [AttributesManager](https://github.com/traccar/traccar/blob/master/src/org/traccar/database/AttributesManager.java), do we really need to copy it?

**Comment by Abyss777** [comment]
I'm not sure... May be it was some reinsurance.
Briefly do not see any problems in `ComputedAttributesHandler`

**Comment by tananaev** [comment]
Thanks. I have create a branch for the implementation:

https://github.com/traccar/traccar/tree/jcache

I suspect there will be a lot of things to fix, so didn't want to commit anything to master yet.

**Comment by Igor-Lopes** [comment]
Adding support for distributed cache will be very helpful, since the current cache system doesn't sync between traccar instances, as discussed on [https://www.traccar.org/forums/topic/high-availability-or-clustering-of-traccar-server/](url)

**Comment by jaimzj** [comment]
Noticed the branch has been removed temporarily? Would you be willing to share more info or timeline on this @tananaev ?

**Comment by tananaev** [comment]
One company sponsored the work, but at one point they decided they want the result to be private, so I had to remove the branch from here. Not sure when it will be available as open source, but it will be eventually.

**Comment by jaimzj** [comment]
Thank you for the response Anton, It would be amazing to have this feature someday soon 👍

**Comment by tananaev** [comment]
I don't think it will be very soon because I would need to re-implement the same work again, but at least I know what and how to do it now.

**Comment by Igor-Lopes** [comment]
Hello, are there any updates on this feature ?

**Comment by mxplusb** [comment]
If @tananaev can provide some guidance on where to find the in-memory cache implementation, I might be able to take a stab at it. I need a distributed deployment of Traccar for resiliency purposes (not worried about scaling)

**Comment by tananaev** [comment]
You can check `jcache` branch.

**Comment by tananaev** [comment]
We completely reimplement caching and added support for synchronization between servers, so horizontal scaling should be supported now. There are some things that are still in progress (commands and status), but most of the other functionality should work correctly.
